dart/lib/src/open_location_code.dart

```diff
@@ -260,21 +328,60 @@ CodeArea decode(String code) {
   code = code.replaceAll(separator, '');
   code = code.replaceAll(RegExp('$padding+'), '');
   code = code.toUpperCase();
-  // Decode the lat/lng pair component.
-  var codeArea =
-      decodePairs(code.substring(0, min(code.length, pairCodeLength)));
-  // If there is a grid refinement component, decode that.
-  if (code.length <= pairCodeLength) {
-    return codeArea;
+  // Initialise the values for each section. We work them out as integers and
+  // convert them to floats at the end.
+  var normalLat = -latitudeMax * pairPrecision;
+  var normalLng = -longitudeMax * pairPrecision;
+  var gridLat = 0;
+  var gridLng = 0;
+  // How many digits do we have to process?
+  var digits = min(code.length, pairCodeLength);
+  // Define the place value for the most significant pair.
+  var pv = pairFirstPlaceValue;
+  // Decode the paired digits.
+  for (var i = 0; i < digits; i += 2) {
+    normalLat += codeAlphabet.indexOf(code[i]) * pv;
+    normalLng += codeAlphabet.indexOf(code[i + 1]) * pv;
+    if (i < digits - 2) {
+      pv = pv ~/ encodingBase;
+    }
   }
-  var gridArea = decodeGrid(
-      code.substring(pairCodeLength, min(code.length, maxDigitCount)));
+  // Convert the place value to a float in degrees.
+  var latPrecision = pv / pairPrecision;
+  var lngPrecision = pv / pairPrecision;
+  // Process any extra precision digits.
+  if (code.length > pairCodeLength) {
+    // Initialise the place values for the grid.
+    var rowpv = gridLatFirstPlaceValue;
+    var colpv = gridLngFirstPlaceValue;
+    // How many digits do we have to process?
+    digits = min(code.length, maxDigitCount);
+    for (var i = pairCodeLength; i < digits; i++) {
+      var digitVal = codeAlphabet.indexOf(code[i]);
+      var row = digitVal ~/ gridColumns;
+      var col = digitVal % gridColumns;
+      gridLat += row * rowpv;
+      gridLng += col * colpv;
+      if (i < digits - 1) {
+        rowpv = rowpv ~/ gridRows;
+        colpv = colpv ~/ gridColumns;
+      }
+    }
+    // Adjust the precisions from the integer values to degrees.
+    latPrecision = rowpv / finalLatPrecision;
+    lngPrecision = colpv / finalLngPrecision;
+  }
+  // Merge the values from the normal and extra precision parts of the code.
+  var lat = normalLat / pairPrecision + gridLat / finalLatPrecision;
+  var lng = normalLng / pairPrecision + gridLng / finalLngPrecision;
+  // Multiple values by 1e14, round and then divide. This reduces errors due
```

Comment:
Nit: Multiply.

Also, any particular reason for 1e14? We use 1e6 in encode above.

Reply:
Turns out I can just remove that.